docs: add Google Cloud Functions example to Serverless Documentation (#3445)
fastify/fastify: docs/Serverless.md

- [Google Cloud Run](#google-cloud-run)
- [Netlify Lambda](#netlify-lambda)
- [Vercel](#vercel)

> - [Google Cloud Functions](#google-cloud-functions)

Reviewer: How are the two Google things different? Regardless, let's alphabetize the list (and order in documentation).

Author: I agree. Thank you @jsumners

- API Gateway does not support streams yet, so you are not able to handle [streams](https://www.fastify.io/docs/latest/Reply/#streams).
- API Gateway has a timeout of 29 seconds, so it is important to provide a reply during this time.

## Google Cloud Functions

### Creation of Fastify instance
```js
const fastify = require("fastify")({
  logger: true // you can also define the level passing an object configuration to logger: {level: 'debug'}
});
```

> ### Add Custom `contentTypeParser` to Fastyfy instance

Reviewer: ```suggestion
### Add Custom `contentTypeParser` to Fastify instance
```

> As explained [here](https://github.com/fastify/fastify/issues/946#issuecomment-766319521), since the Google Cloud Functions platform parses the body of the request before it arrives into Fastify instance, troubling the body request in case of `POST` and `PATCH` methods, you need to add a custom [`ContentTypeParser`](https://www.fastify.io/docs/latest/ContentTypeParser/) to mitigate this behavior.

Reviewer: ```suggestion
As explained [in issue #946](https://github.com/fastify/fastify/issues/946#issuecomment-766319521), since the Google Cloud Functions platform parses the body of the request before it arrives into Fastify instance, troubling the body request in case of `POST` and `PATCH` methods, you need to add a custom [`ContentTypeParser`](https://www.fastify.io/docs/latest/ContentTypeParser/) to mitigate this behavior.
```

Author: Cleanest... Thank you 👍

```js
fastify.addContentTypeParser('application/json', {}, (req, body, done) => {
  done(null, body.body);
});
```

### Define your endpoint (examples)

> A simple `GET` endpoint

Reviewer: ```suggestion
A simple `GET` endpoint:
```

```js
fastify.get('/', async (request, reply) => {
  reply.send({message: 'Hello World!'})
})
```

> Or a more complete `POST` endpoint with schema validation 

Reviewer: ```suggestion
Or a more complete `POST` endpoint with schema validation: 
```

> ```JS

Reviewer: For consistency:

```suggestion
```js
```

fastify.route({
  method: 'POST',
  url: '/hello',
  schema: {
    body: {
      type: 'object',
      properties: {
        name: { type: 'string'}
      },
      required: ['name']
    },
    response: {
      200: {
        type: 'object', 
        properties: {
          message: {type: 'string'}
        }
      }
    },
  },
  handler: async (request, reply) => {
    const { name } = request.body;
    reply.code(200).send({
      message: `Hello ${name}!`
    })
  }
})
```

### Implement and export the function

> Final step, implement the function to handle the request and passing it to Fastify by emitting `request` event to `fastify.server`

Reviewer: ```suggestion
Final step, implement the function to handle the request and pass it to Fastify by emitting `request` event to `fastify.server`:
```

```js
const fastifyFunction = async (request, reply) => {
  await fastify.ready();
  fastify.server.emit('request', request, reply)
}

export.fastifyFunction = fastifyFunction;
```

### Local test

Install [Google Functions Framework for Node.js](https://github.com/GoogleCloudPlatform/functions-framework-nodejs).

> You can install it globally

Reviewer: ```suggestion
You can install it globally:
```

```bash
npm i -g @google-cloud/functions-framework
```

> or as a development library.

Reviewer: ```suggestion
Or as a development library:
```

```bash
npm i --save-dev @google-cloud/functions-framework
```

> Than you can run your function locally with Functions Framework

Reviewer: ```suggestion
Than you can run your function locally with Functions Framework:
```

``` bash
npx @google-cloud/functions-framework --target=fastifyFunction
```

> or add this command to your `package.json` scripts 

Reviewer: ```suggestion
Or add this command to your `package.json` scripts:
```

    app.server.emit('request', req, res);
}
```

## Google Cloud Functions

### Creation of Fastify instance
```js
const fastify = require("fastify")({
  logger: true // you can also define the level passing an object configuration to logger: {level: 'debug'}
});
```

### Add Custom `contentTypeParser` to Fastyfy instance

As explained [here](https://github.com/fastify/fastify/issues/946#issuecomment-766319521), since the Google Cloud Functions platform parses the body of the request before it arrives into Fastify instance, troubling the body request in case of `POST` and `PATCH` methods, you need to add a custom [`ContentTypeParser`](https://www.fastify.io/docs/latest/ContentTypeParser/) to mitigate this behavior.

```js
fastify.addContentTypeParser('application/json', {}, (req, body, done) => {
  done(null, body.body);
});
```

> ### Defines your endpoint (examples)

Reviewer 2: ```suggestion
### Define your endpoint (examples)
```

A simple `GET` endpoint
```js
fastify.get('/', async (request, reply) => {

>   replay.send({message: 'Hello World!'})

Reviewer 2: ```suggestion
  reply.send({message: 'Hello World!'})
```

Author: Awesome! Thank you @matthyk

    app.server.emit('request', req, res);
}
```

## Google Cloud Functions

> 
> You can easly use Fastify inside your Google Cloud Functions.

Reviewer 3: ```suggestion
```

Redundant line, also a [tad condescending](https://github.com/fastify/fastify/blob/main/docs/Style-Guide.md#avoid-using-condescending-terms).

### Creation of Fastify instance
```js
const fastify = require("fastify")({
  logger: true // you can also define the level passing an object configuration to logger: {level: 'debug'}
});
```

### Add Custom `contentTypeParser` to Fastyfy instance

> As explained [here](https://github.com/fastify/fastify/issues/946#issuecomment-766319521), since Google Cloud Functions platform perform the parsing of the body request before it arrives into Fastify instance, troubling the body request in case of `POST` and `PATCH` methods, you need to add a custom [`ContentTypeParser`](https://www.fastify.io/docs/latest/ContentTypeParser/) to mitigate this behavior.

Reviewer 3: ```suggestion
As explained [here](https://github.com/fastify/fastify/issues/946#issuecomment-766319521), since the Google Cloud Functions platform parses the body of the request before it arrives into Fastify instance, troubling the body request in case of `POST` and `PATCH` methods, you need to add a custom [`ContentTypeParser`](https://www.fastify.io/docs/latest/ContentTypeParser/) to mitigate this behavior.
```

Author: Thank you @Fdawgs 
Now is much better... :-)

    app.server.emit('request', req, res);
}
```

## Google Cloud Functions

> You can run Fastify as a Google Cloud Functions using [`serverFactory`](https://www.fastify.io/docs/v1.13.x/Server/#serverfactory) option to pass the cloud function HTTP request to Fastify `handler`.

Reviewer 4: We should link to the latest docs

### Creation of Fastify instance with `serverFactory`
```js
const fastify = require("fastify")({
  serverFactory: (handler, opts) => {
    const server = require("http").createServer((req, res) => {
      handler(req, res);
    });
    return server;
  },

> });

Reviewer 5: I don't think this part is strictly needed. Is it?

Author: I took the example code from [serverFactory documentation](https://www.fastify.io/docs/latest/Server/#serverfactory) to easily suggest a starting point.

I can remove it, leaving only the link to the documentation. Let me know what you consider to be the best option.

Alternatively, I could also change the function to return directly the server like this

```js
const fastify = require("fastify")({
  serverFactory: (handler, opts) => {
    return require("http").createServer((req, res) => {
      handler(req, res);
    });
  },
});
```

Reviewer 6: The original post need this part because the `handler` need to be extracted for `firebase-functions`.

If you plan to advice people using `fastify.server.emit('request', request, reply)`, then a custom `serverFactory` actually do nothing but confusion to user why it is needed.

Author: Thank you @climba03003. You are right. The first version of my code uses a handler created with `serverFactory`. 

Then I found an example with `fastify.server.emit('request', request, reply)` and I found it more elegant. :-)

This reveals my not complete understanding of `serverFactory` function... :-D 

In any case... I've tested the script without the `serverFactory` option and it works.

What do you think is a better approach? For example for performance, if it is relevant, is it better to use `serverFactory` or `fastify.server.emit`?

Reviewer 6: Keep the usage of `fastify.server.emit`, let `fastify` handle the `http` creation.

Author: Awesome. 

I agree the final result is much more elegant and easy to understand.

Can I ask you where could I find documentation about events that could be emitted to `fastify.server`? Just for my study :-)

Reviewer 6: `fastify.server` is `http.Server`. So, check the official document for `nodejs`.
https://nodejs.org/dist/latest-v16.x/docs/api/http.html#event-request

```

> ### Add Custom `contentTypeParse` to Fastyfy instance

Reviewer 7: ```suggestion
### Add Custom `contentTypeParser` to Fastify instance
```

> As explained [here](https://github.com/fastify/fastify/issues/946#issuecomment-766319521), since Google Cloud Functions platform perform the parsing of the body request before it arrives into Fastify instance, troubling the body request in case of `POST` and `PATCH` methods, you need to add a custom [`ContentTypeParser`](https://www.fastify.io/docs/v1.13.x/ContentTypeParser/) to mitigate this behavior.

Reviewer 4: Old doc link

Author: Updated link to latest path. Thank you
